File: leet-code/averageOfLevels.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

function TreeNode (val) {
  this.val = val
  this.left = this.right = null
}
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function (root) {
  let queue = []
  let result = []
  queue.push(root)
  while (queue.length > 0) {
    let levelSize = queue.length
    let tempVal = 0
    for (let i = 0; i < levelSize; i++) {
      // 出栈
      let tempNode = queue.shift()
      tempVal += tempNode.val
      if (tempNode.left != null) {
        queue.push(tempNode.left)
      }
      if (tempNode.right != null) {
        queue.push(tempNode.right)
      }
    }
    result.push(tempVal/levelSize)
  }
  return result
}

File: leet-code/averageOfLevels.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
	var result []float64
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		levelSize := len(queue)
		temp := 0
		for i := 0; i < levelSize; i++ {
			node := queue[i]
			temp += node.Val

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		queue = queue[levelSize:]
		result = append(result, float64(temp)/float64(levelSize))
	}
	return result
}
